File: wear/src/main/java/com/codegy/aerlink/connection/ConnectionManager.java
package com.codegy.aerlink.connection;

import android.bluetooth.*;
import android.content.Context;
import android.os.Handler;
import android.util.Log;
import com.codegy.aerlink.connection.characteristic.CharacteristicIdentifier;
import com.codegy.aerlink.connection.characteristic.CharacteristicSubscriber;
import com.codegy.aerlink.connection.characteristic.CharacteristicSubscriberThread;
import com.codegy.aerlink.connection.command.Command;
import com.codegy.aerlink.connection.command.CommandHandler;
import com.codegy.aerlink.connection.command.CommandQueueThread;

import java.lang.reflect.Method;
import java.util.*;

import static android.content.ContentValues.TAG;

/**
 * Created by Guiye on 18/5/15.
 */
public class ConnectionManager implements CharacteristicSubscriber, CommandHandler {

    private static final String LOG_TAG = ConnectionManager.class.getSimpleName();

    private static final String DESCRIPTOR_CONFIG = "00002902-0000-1000-8000-00805f9b34fb";

    public interface Callback {
        void onConnectionStateChange(ConnectionState state);
        void onReadyToSubscribe(BluetoothGatt bluetoothGatt);
        void onCharacteristicChanged(BluetoothGattCharacteristic characteristic);
    }

    private Context mContext;
    private Callback mCallback;
    private BluetoothDevice mDevice;

    private BluetoothAdapter mBluetoothAdapter;

    private int mBondsFailed = 0;
    private int mConnectionsFailed = 0;

    private CharacteristicSubscriberThread subscriberThread;
    private CommandQueueThread commandQueue;


    public ConnectionManager(Context context, Callback callback, BluetoothAdapter bluetoothAdapter) {
        this.mContext = context;
        this.mCallback = callback;
        this.mBluetoothAdapter = bluetoothAdapter;

        this.commandQueue = new CommandQueueThread(this);
        this.commandQueue.start();
    }


    public void close() {
        // Check if already closed
        if (mCallback == null) {
            return;
        }

        Log.i(LOG_TAG, "Close");
        mCallback = null;

        disconnectDevice();

        if (subscriberThread != null) {
            subscriberThread.kill();
            subscriberThread = null;
        }

        if (commandQueue != null) {
            commandQueue.kill();
            commandQueue = null;
        }
    }

    private void reset() {
        // Check if already closed
        if (mCallback == null) {
            return;
        }

        if (subscriberThread != null) {
            subscriberThread.kill();
            subscriberThread = null;
        }

        if (commandQueue != null) {
            commandQueue.clear();

            commandQueue.setReady(false);
        }

        // mCallback.onConnectionStateChange(ConnectionState.Disconnected);
    }

    public synchronized void connectToDevice(final BluetoothDevice device) {
        if (mBluetoothGatt != null) {
            return;
        }

        mDevice = device;

        if (subscriberThread == null) {
            subscriberThread = new CharacteristicSubscriberThread(this);
            subscriberThread.start();
        }
        else {
            subscriberThread.reset();
        }

        Handler handler = new Handler(mContext.getMainLooper());
        handler.post(new Runnable() {
            @Override
            public void run() {
                mBluetoothGatt = device.connectGatt(mContext, false, mBluetoothGattCallback);

                Log.i(LOG_TAG, "Connecting...: " + device.getName());
                mCallback.onConnectionStateChange(ConnectionState.Connecting);
            }
        });
    }

    private synchronized void disconnectDevice() {
        try {
            if (mBluetoothGatt != null) {
                mBluetoothGatt.close();
            }
        }
        catch (Exception e) {
            e.printStackTrace();
        }
        finally {
            mBluetoothGatt = null;
        }
    }

    public synchronized void tryToReconnect() {
        if (mCallback == null) {
            return;
        }

        reset();
        disconnectDevice();

        if (mDevice == null) {
            mCallback.onConnectionStateChange(ConnectionState.Disconnected);
            return;
        }

        Log.w(LOG_TAG, "Reconnecting");
        connectToDevice(mDevice);
    }

    private boolean changeGattMtu(int mtu) {
        int retry = 5;
        boolean ok = false;
        while (!ok && retry > 0) {
            ok = mBluetoothGatt.requestMtu(mtu);
            retry--;
        }

        return ok;
    }

    private BluetoothGatt mBluetoothGatt;
    private final BluetoothGattCallback mBluetoothGattCallback = new BluetoothGattCallback() {

        @Override
        public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
            super.onConnectionStateChange(gatt, status, newState);
            Log.d(LOG_TAG, "onConnectionStateChange: " + status + " -> " + newState);


            if (status == BluetoothGatt.GATT_SUCCESS) {
                switch (newState) {
                    case BluetoothProfile.STATE_CONNECTED:
                        if (mCallback != null && gatt == mBluetoothGatt) {
                            gatt.discoverServices();

                            subscriberThread.setDiscovering();
                        }
                        break;
                    case BluetoothProfile.STATE_DISCONNECTED:
                        Log.e(LOG_TAG, "Disconnected");
                        gatt.close();
                        Log.w(LOG_TAG, "Closed");
                        if (mCallback == null || gatt != mBluetoothGatt) {
                            return;
                        }

                        mBluetoothGatt = null;

                        tryToReconnect();
                        break;
                }
            }
            else {
                Log.wtf(LOG_TAG, "ON CONNECTION STATE CHANGED ERROR: " + status);

                gatt.close();
                Log.w(LOG_TAG, "Closed");
                if (mCallback == null || gatt != mBluetoothGatt) {
                    return;
                }

                mBluetoothGatt = null;

                // 8 means time out, don't restart bluetooth, just reconnect when possible
                if (status != 8) {
                    //BluetoothUtils.resetBondedDevices(mBluetoothAdapter);
                    mDevice = null;
                    BluetoothUtils.disableBluetooth(mBluetoothAdapter);
                    return;
                }

                tryToReconnect();
            }
        }

        @Override
        public void onServicesDiscovered(BluetoothGatt gatt, int status) {
            super.onServicesDiscovered(gatt, status);
            Log.d(LOG_TAG, "onServicesDiscovered: " + status);
            // Check if already closed
            if (mCallback == null || gatt != mBluetoothGatt) {
                return;
            }

            if (status == BluetoothGatt.GATT_SUCCESS) {
                if (!changeGattMtu(512)) {
                    mCallback.onReadyToSubscribe(gatt);
                }
            }
        }

        @Override
        public void onDescriptorWrite(BluetoothGatt gatt, BluetoothGattDescriptor descriptor, int status) {
            super.onDescriptorWrite(gatt, descriptor, status);
            // Check if already closed
            if (mCallback == null || gatt != mBluetoothGatt) {
                return;
            }

            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(LOG_TAG, "Descriptor write successful: " + descriptor.getCharacteristic().getUuid().toString());

                subscriberThread.remove();
            }
            else {
                BluetoothDevice device = gatt.getDevice();
                int bondState = device.getBondState();

                // Don't do anything while bonding
                if (bondState != BluetoothDevice.BOND_BONDED) {
                    Log.e(LOG_TAG, "Status: write not permitted");

                    BluetoothUtils.unpairDevice(gatt.getDevice());
                    mDevice = null;
                    tryToReconnect();
                }
            }
        }

        @Override
        public void onCharacteristicWrite(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic, int status) {
            super.onCharacteristicWrite(gatt, characteristic, status);
            // Check if already closed
            if (mCallback == null || gatt != mBluetoothGatt) {
                return;
            }

            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(LOG_TAG, "Characteristic write successful: " + characteristic.getUuid().toString());

                commandQueue.remove();
            }
            else {
                Log.w(LOG_TAG, "Characteristic write error: " + status + " :: " + characteristic.getUuid().toString());

                commandQueue.moveToBack();
            }

            Log.d(LOG_TAG, "Characteristic value: " + new String(characteristic.getValue()));
        }

        @Override
        public void onCharacteristicRead(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic, int status) {
            super.onCharacteristicRead(gatt, characteristic, status);
            Log.d(LOG_TAG, "onCharacteristicRead status:: " + status);
            // Check if already closed
            if (mCallback == null || gatt != mBluetoothGatt) {
                return;
            }

            if (status == BluetoothGatt.GATT_SUCCESS) {
                mCallback.onCharacteristicChanged(characteristic);
            }

            commandQueue.remove();
        }

        @Override
        public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
            super.onCharacteristicChanged(gatt, characteristic);
            Log.d(LOG_TAG, "onCharacteristicChanged: " + characteristic.getUuid().toString());
            // Check if already closed
            if (mCallback == null || gatt != mBluetoothGatt) {
                return;
            }
            
            mCallback.onCharacteristicChanged(characteristic);
        }

        @Override
        public void onMtuChanged(BluetoothGatt gatt, int mtu, int status) {
            super.onMtuChanged(gatt, mtu, status);
            Log.d(LOG_TAG, "onMtuChanged: " + mtu + " status: " + status);
            if (mCallback == null || gatt != mBluetoothGatt) {
                return;
            }

            if (status ==  BluetoothGatt.GATT_SUCCESS) {
                mCallback.onReadyToSubscribe(gatt);
            }
        }
    };

    private BluetoothGattCharacteristic findCharacteristic(UUID serviceUUID, UUID characteristicUUID) {
        if (mBluetoothGatt == null) {
            return null;
        }

        BluetoothGattService service = mBluetoothGatt.getService(serviceUUID);
        if (service == null) {
            Log.e(LOG_TAG, "Service unavailable");
            return null;
        }

        BluetoothGattCharacteristic characteristic = service.getCharacteristic(characteristicUUID);
        if (characteristic == null) {
            Log.e(LOG_TAG, "Characteristic unavailable");
            return null;
        }

        return characteristic;
    }

    // region Commands

    public void addCommandToQueue(Command command) {
        commandQueue.put(command);
    }

    @Override
    public void handleCommand(Command command) {
        if (mBluetoothGatt == null) {
            return;
        }

        final Command finalCommand = command;

        Handler handler = new Handler(mContext.getMainLooper());
        handler.post(new Runnable() {
            @Override
            public void run() {
                try {
                    BluetoothGattCharacteristic characteristic = findCharacteristic(finalCommand.getServiceUUID(), finalCommand.getCharacteristic());
                    if (characteristic == null) {
                        return;
                    }

                    if (finalCommand.isWriteCommand()) {
                        // not being used
                        // mBluetoothGattCharacteristic.setWriteType(command.getWriteType());

                        characteristic.setValue(finalCommand.getPacket());
                        characteristic.setWriteType(BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT);
                        boolean result = mBluetoothGatt.writeCharacteristic(characteristic);
                        Log.d(LOG_TAG, "Started writing command: " + result);
                    }
                    else {
                        boolean result = mBluetoothGatt.readCharacteristic(characteristic);
                        Log.d(LOG_TAG, "Started reading command: " + result);
                    }
                }
                catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
    }

    // endregion


    //region Subscribe requests

    public void addSubscribeRequests(Queue<CharacteristicIdentifier> requests) {
        subscriberThread.setSubscribeRequests(requests);
    }

    @Override
    public void subscribeCharacteristic(CharacteristicIdentifier characteristicIdentifier) {
        if (mBluetoothGatt == null) {
            return;
        }

        final CharacteristicIdentifier finalCharacteristicIdentifier = characteristicIdentifier;

        Handler handler = new Handler(mContext.getMainLooper());
        handler.post(new Runnable() {
            @Override
            public void run() {
                try {
                    BluetoothGattCharacteristic characteristic = findCharacteristic(finalCharacteristicIdentifier.getServiceUUID(), finalCharacteristicIdentifier.getCharacteristicUUID());
                    if (characteristic == null) {
                        return;
                    }

                    mBluetoothGatt.setCharacteristicNotification(characteristic, true);

                    BluetoothGattDescriptor descriptor = characteristic.getDescriptor(UUID.fromString(DESCRIPTOR_CONFIG));
                    if (descriptor == null) {
                        Log.e(LOG_TAG, "Descriptor unavailable to write");
                        return;
                    }

                    descriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
                    //descriptor.setValue(BluetoothGattDescriptor.ENABLE_INDICATION_VALUE);
                    boolean result = mBluetoothGatt.writeDescriptor(descriptor);
                    Log.d(LOG_TAG, "Started writing descriptor: " + result);
                }
                catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
    }

    //endregion

    @Override
    public void onConnectionFailed() {
        Log.w(LOG_TAG, "Connecting timed out");
        if (mCallback == null) {
            return;
        }

        if (mBluetoothGatt != null && mBluetoothGatt.getDevice() != null) {
            BluetoothDevice device = mBluetoothGatt.getDevice();
            int bondState = device.getBondState();

            // Don't do anything while bonding
            if (bondState == BluetoothDevice.BOND_BONDING && mBondsFailed < 30) {
                Log.w(LOG_TAG, "Waiting for bond...");
                mBondsFailed++;

                subscriberThread.setConnecting();
            }
            else {
                mBondsFailed = 0;
                mConnectionsFailed++;

                BluetoothUtils.restartBluetooth(mBluetoothAdapter);

                if (mConnectionsFailed%3 == 0) {
                    mDevice = null;
                }

                tryToReconnect();
            }
        }
        else {
            Log.w(LOG_TAG, "Start scanning again");

            mDevice = null;

            tryToReconnect();
        }
    }

    @Override
    public void onSubscribingFailed() {
        Log.w(LOG_TAG, "Subscribing timed out");
        if (mCallback == null) {
            return;
        }

        if (mBluetoothGatt != null && mBluetoothGatt.getDevice() != null) {
            BluetoothDevice device = mBluetoothGatt.getDevice();
            int bondState = device.getBondState();

            // Don't do anything while bonding
            if (bondState == BluetoothDevice.BOND_BONDING && mBondsFailed < 30) {
                Log.w(LOG_TAG, "Waiting for bond...");
                mBondsFailed++;

                subscriberThread.setBonding();
            }
            else {
                mBondsFailed = 0;
                mConnectionsFailed++;

                if (mConnectionsFailed >= 10) {
                    mConnectionsFailed = 0;

                    // Last resort to prepare for a new connection
                    // Reset bonded devices
                    BluetoothUtils.resetBondedDevices(mBluetoothAdapter);
                }

                BluetoothUtils.restartBluetooth(mBluetoothAdapter);

                if (mConnectionsFailed%3 == 0) {
                    mDevice = null;
                }

                tryToReconnect();
            }
        }
        else {
            Log.w(LOG_TAG, "Start scanning again");

            mDevice = null;

            tryToReconnect();
        }
    }

    @Override
    public void onConnectionReady() {
        Log.i(LOG_TAG, "Ready");

        mCallback.onConnectionStateChange(ConnectionState.Ready);

        mBondsFailed = 0;
        mConnectionsFailed = 0;

        commandQueue.setReady(true);

        // We are done with this thread, we don't need it until disconnection
        subscriberThread = null;
    }

}

File: wear/src/main/java/com/codegy/aerlink/connection/ConnectionManager.kt
package com.codegy.aerlink.connection

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.content.Context
import android.os.Looper
import android.util.Log
import com.codegy.aerlink.extensions.getCharacteristic
import com.codegy.aerlink.extensions.stringFromConnectionError
import com.codegy.aerlink.extensions.stringFromStatus
import com.codegy.aerlink.extensions.subscribeToCharacteristic
import com.codegy.aerlink.extensions.unpair
import com.codegy.aerlink.utils.ScheduledTask
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicBoolean

class ConnectionManager(val device: BluetoothDevice, private val context: Context, var callback: Callback?, bluetoothManager: BluetoothManager) {
    interface Callback {
        fun onReadyToSubscribe(connectionManager: ConnectionManager): List<CharacteristicIdentifier>
        fun onConnectionReady(connectionManager: ConnectionManager)
        fun onDisconnected(connectionManager: ConnectionManager)
        fun onConnectionError(connectionManager: ConnectionManager)
        fun onBondingRequired(connectionManager: ConnectionManager, device: BluetoothDevice)
        fun onCharacteristicChanged(connectionManager: ConnectionManager, characteristic: BluetoothGattCharacteristic)
    }

    private var atomicRunning = AtomicBoolean(true)
    var isRunning: Boolean
        get() = atomicRunning.get()
        private set(value) = atomicRunning.set(value)

    private val lock = Object()
    private val timeoutController: ScheduledTask = ScheduledTask(Looper.getMainLooper())
    private val characteristicsToSubscribe: Queue<CharacteristicIdentifier> = ConcurrentLinkedQueue()
    private var currentCommand: Command? = null
    private val commands: Queue<Command> = ConcurrentLinkedQueue()

    private val bluetoothAdapter: BluetoothAdapter = bluetoothManager.adapter
    private var bluetoothGatt: BluetoothGatt? = null
    private val bluetoothGattCallback = object : BluetoothGattCallback() {

        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            super.onConnectionStateChange(gatt, status, newState)

            timeoutController.cancel()
            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(LOG_TAG, "Connection state changed :: State: $newState")
                when (newState) {
                    BluetoothProfile.STATE_CONNECTED -> {
                        Log.i(LOG_TAG, "Connected")
                        timeoutController.schedule(OPERATION_DELAY) {
                            if (!isRunning) { return@schedule }
                            requestMtu(gatt)
                        }
                    }
                    BluetoothProfile.STATE_DISCONNECTED -> {
                        Log.i(LOG_TAG, "Disconnected")
                        gatt.close()
                        bluetoothGatt = null

                        callback?.onDisconnected(this@ConnectionManager)
                    }
                }
            } else {
                Log.e(LOG_TAG, "ERROR: Connection state changed :: Status: ${stringFromConnectionError(status)}")
                onConnectionError()
            }
        }

        fun requestMtu(gatt: BluetoothGatt, numberOfTries: Int = 0) {
            val success = gatt.requestMtu(DESIRED_MTU)
            if (!success) {
                if (numberOfTries < 3) {
                    Log.w(LOG_TAG, "Error requesting MTU change: $numberOfTries")
                    requestMtu(gatt, numberOfTries + 1)
                } else {
                    Log.e(LOG_TAG, "Too many errors requesting MTU change, will try to discover services")
                    discoverServices(gatt)
                }
                return
            }
            Log.d(LOG_TAG, "Success requesting MTU change: $DESIRED_MTU")

            timeoutController.schedule(MTU_CHANGE_TIMEOUT) {
                if (!isRunning) { return@schedule }
                discoverServices(gatt)
            }
        }

        override fun onMtuChanged(gatt: BluetoothGatt, mtu: Int, status: Int) {
            super.onMtuChanged(gatt, mtu, status)

            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(LOG_TAG, "MTU changed :: MTU: $mtu")
            } else {
                Log.e(LOG_TAG, "ERROR: MTU changed :: Status: ${stringFromStatus(status)}")
            }

            timeoutController.cancel()
            // Continue with connection even if the MTU change failed
            timeoutController.schedule(OPERATION_DELAY) {
                if (!isRunning) { return@schedule }
                discoverServices(gatt)
            }
        }

        fun discoverServices(gatt: BluetoothGatt, numberOfTries: Int = 0) {
            val success = gatt.discoverServices()
            if (!success) {
                if (numberOfTries < 3) {
                    Log.w(LOG_TAG, "Error requesting service discovery: $numberOfTries")
                    discoverServices(gatt, numberOfTries + 1)
                } else {
                    Log.e(LOG_TAG, "Too many errors requesting service discovery, giving up on connection")
                    onConnectionError()
                }
                return
            }
            Log.d(LOG_TAG, "Success requesting service discovery: $DESIRED_MTU")

            timeoutController.schedule(SERVICE_DISCOVERY_TIMEOUT) {
                if (!isRunning) { return@schedule }
                onConnectionTimedOut()
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            super.onServicesDiscovered(gatt, status)

            timeoutController.cancel()
            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(LOG_TAG, "Services Discovered: ${gatt.services}")
                val characteristics = callback?.onReadyToSubscribe(this@ConnectionManager)
                if (characteristics?.isNotEmpty() == true) {
                    characteristicsToSubscribe.addAll(characteristics)
                    timeoutController.schedule(OPERATION_DELAY) {
                        if (!isRunning) { return@schedule }
                        checkCharacteristicsToSubscribe()
                    }
                } else {
                    Log.wtf(LOG_TAG, "No services available, giving up on connection")
                    onConnectionError()
                }
            } else {
                Log.e(LOG_TAG, "ERROR: Services Discovered :: Status: ${stringFromStatus(status)}")
                onConnectionError()
            }
        }

        fun checkCharacteristicsToSubscribe() {
            val characteristic = characteristicsToSubscribe.poll() ?: run {
                connectionsFailed = 0
                callback?.onConnectionReady(this@ConnectionManager)
                return
            }
            subscribeToCharacteristic(characteristic)
        }

        fun subscribeToCharacteristic(characteristicIdentifier: CharacteristicIdentifier, numberOfTries: Int = 0) {
            val success = bluetoothGatt?.subscribeToCharacteristic(characteristicIdentifier) == true
            if (!success) {
                if (numberOfTries < 3) {
                    subscribeToCharacteristic(characteristicIdentifier, numberOfTries + 1)
                } else {
                    onConnectionError()
                }
                return
            }

            timeoutController.schedule(SERVICE_SUBSCRIPTION_TIMEOUT) {
                if (!isRunning) { return@schedule }
                onConnectionTimedOut()
            }
        }

        override fun onDescriptorWrite(gatt: BluetoothGatt, descriptor: BluetoothGattDescriptor, status: Int) {
            super.onDescriptorWrite(gatt, descriptor, status)

            timeoutController.cancel()
            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(LOG_TAG, "Descriptor write: ${descriptor.characteristic.uuid}")

                timeoutController.schedule(OPERATION_DELAY) {
                    if (!isRunning) { return@schedule }
                    checkCharacteristicsToSubscribe()
                }
            } else {
                Log.e(LOG_TAG, "ERROR: Descriptor write :: Status: ${stringFromStatus(status)}")
                Log.d(LOG_TAG, "Descriptor: ${descriptor.characteristic.uuid}")
                onConnectionError()
            }
        }

        fun checkCommands() {
            if (currentCommand != null) {
                return
            }
            commands.poll()?.let {
                currentCommand = it
                handleCommand(it)
            }
        }

        fun handleCommand(command: Command, numberOfTries: Int = 0) {
            val characteristic = bluetoothGatt?.getCharacteristic(command.serviceUUID, command.characteristicUUID) ?: return

            val success: Boolean
            if (command.isWriteCommand) {
                characteristic.value = command.packet
                characteristic.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
                success = bluetoothGatt?.writeCharacteristic(characteristic) == true
                Log.d(LOG_TAG, "Started writing command: $success")
            } else {
                success = bluetoothGatt?.readCharacteristic(characteristic) == true
                Log.d(LOG_TAG, "Started reading command: $success")
            }

            if (!success) {
                if (numberOfTries < 3) {
                    handleCommand(command, numberOfTries + 1)
                } else {
                    onConnectionError()
                }
                return
            }

            timeoutController.schedule(COMMAND_TIMEOUT) {
                if (!isRunning) { return@schedule }
                onConnectionError()
            }
        }

        override fun onCharacteristicWrite(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
            super.onCharacteristicWrite(gatt, characteristic, status)

            timeoutController.cancel()
            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(LOG_TAG, "Characteristic write: ${characteristic.uuid}")

                currentCommand?.completeWithSuccess()
            } else {
                Log.e(LOG_TAG, "ERROR: Characteristic write :: Status: ${stringFromStatus(status)}")
                Log.d(LOG_TAG, "Characteristic: ${characteristic.uuid}")

                currentCommand?.let {
                    it.completeWithFailure()
                    if (it.shouldRetryAgain()) {
                        commands.add(it)
                    }
                }
            }
            Log.d(LOG_TAG, "Characteristic value: ${characteristic.value}")

            currentCommand = null
            timeoutController.schedule(OPERATION_DELAY) {
                if (!isRunning) { return@schedule }
                checkCommands()
            }
        }

        override fun onCharacteristicRead(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
            super.onCharacteristicRead(gatt, characteristic, status)

            timeoutController.cancel()
            if (status == BluetoothGatt.GATT_SUCCESS) {
                Log.d(LOG_TAG, "Characteristic read: ${characteristic.uuid}")
                callback?.onCharacteristicChanged(this@ConnectionManager, characteristic)

                currentCommand?.completeWithSuccess()
            } else {
                Log.e(LOG_TAG, "ERROR: Characteristic read :: Status: ${stringFromStatus(status)}")
                Log.d(LOG_TAG, "Characteristic: ${characteristic.uuid}")

                currentCommand?.let {
                    it.completeWithFailure()
                    if (it.shouldRetryAgain()) {
                        commands.add(it)
                    }
                }
            }
            Log.d(LOG_TAG, "Characteristic value: ${characteristic.value}")

            currentCommand = null
            timeoutController.schedule(OPERATION_DELAY) {
                if (!isRunning) { return@schedule }
                checkCommands()
            }
        }

        override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
            super.onCharacteristicChanged(gatt, characteristic)
            Log.d(LOG_TAG, "Characteristic changed: ${characteristic.uuid}")

            callback?.onCharacteristicChanged(this@ConnectionManager, characteristic)
        }
    }

    fun connect(autoConnect: Boolean = false) {
        if (device.bondState != BluetoothDevice.BOND_BONDED) {
            Log.w(LOG_TAG, "Trying to connect to an unbonded device")
            callback?.onBondingRequired(this, device)
            return
        }

        synchronized(lock) {
            characteristicsToSubscribe.clear()
            bluetoothGatt?.close()

            Log.i(LOG_TAG, "Connecting...: " + device.name)
            bluetoothGatt = device.connectGatt(context, autoConnect, bluetoothGattCallback)
            if (!autoConnect) {
                timeoutController.schedule(CONNECTION_TIMEOUT) {
                    if (!isRunning) { return@schedule }
                    onConnectionTimedOut()
                }
            }
        }
    }

    fun disconnect() {
        synchronized(lock) {
            callback?.onDisconnected(this)

            try {
                bluetoothGatt?.close()
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                bluetoothGatt = null
            }
        }
    }

    fun checkServiceAvailability(serviceUuid: UUID): Boolean {
        return bluetoothGatt?.getService(serviceUuid) != null
    }

    fun handleCommand(command: Command) {
        if (!isRunning || bluetoothGatt == null) {
            return
        }

        commands.add(command)
        bluetoothGattCallback.checkCommands()
    }

    private fun onConnectionError(unpairDevice: Boolean = false) {
        bluetoothGatt?.close()
        Log.w(LOG_TAG, "Closed BluetoothGatt")
        bluetoothGatt = null

        if (unpairDevice) {
            device.unpair()
            connectionsFailed = 0
        }

        bluetoothAdapter.disable()

        callback?.onConnectionError(this@ConnectionManager)
    }

    private fun onConnectionTimedOut() {
        Log.w(LOG_TAG, "Connecting timed out")
        connectionsFailed++
        if (connectionsFailed > 0 && connectionsFailed % 3 == 0) {
            onConnectionError(connectionsFailed % 9 == 0)
        } else {
            disconnect()
        }
    }

    fun close() {
        // Check if already closed
        if (!isRunning) {
            return
        }

        isRunning = false
        callback = null
        disconnect()
        Log.i(LOG_TAG, "ConnectionManager Closed")
    }

    companion object {
        private var connectionsFailed = 0
        private val LOG_TAG = ConnectionManager::class.java.simpleName
        private const val DESIRED_MTU: Int = 512
        private const val OPERATION_DELAY: Long = 180
        private const val CONNECTION_TIMEOUT: Long = 5000
        private const val MTU_CHANGE_TIMEOUT: Long = 1000
        private const val SERVICE_DISCOVERY_TIMEOUT: Long = 2000
        private const val SERVICE_SUBSCRIPTION_TIMEOUT: Long = 3000
        private const val COMMAND_TIMEOUT: Long = 2000
    }
}
